kafkaesk/retry.py

```diff
@@ -0,0 +1,157 @@
+from .exceptions import UnhandledMessage
+from .metrics import MESSAGE_FAILED
+from .metrics import MESSAGE_REQUEUED
+from abc import ABC
+from aiokafka.structs import ConsumerRecord
+from pydantic import BaseModel
+from typing import Callable
+from typing import TYPE_CHECKING
+
+import datetime
+
+if TYPE_CHECKING:
+    from .app import Application
+    from .app import Subscription
+
+
+class RetryInfo(BaseModel):
+    retry_count: int
+    retry_delay: int = 0
+    retry_timestamp: datetime.datetime
+    publish_timestamp: datetime.datetime
+    publish_topic: str
+    error: str
+
+
+class RetryMessage(BaseModel):
+    retry_info: RetryInfo
+    original_record: ConsumerRecord
+
+    class Config:
+        arbitrary_types_allowed = True
+
+
+class FailureInfo(BaseModel):
+    retry_count: int
+    failure_timestamp: datetime.datetime
+    publish_topic: str
+    error: str
+
+
+class FailureMessage(BaseModel):
+    failure_info: FailureInfo
+    original_record: ConsumerRecord
+
+    class Config:
+        arbitrary_types_allowed = True
+
+
+class RetryPolicy(ABC):
+    def __init__(self) -> None:
+        self._initialized = False
+
+    async def initialize(self, app: "Application", subscription: "Subscription") -> None:
+
+        self._app = app
+        self._subscription = subscription
+
+        self._initialized = True
+
+    async def finalize(self) -> None:
+        self._initialized = False
+
+    async def __call__(self, record: ConsumerRecord, error: Exception) -> None:
+        if self._initialized is not True:
+            raise RuntimeError("RetryPolicy is not initialized")
+
+        if self._should_retry(record, error):
+            return await self._handle_retry(record, error)
+        else:
+            return await self._handle_failure(record, error)
+
+    def _should_retry(self, record: ConsumerRecord, error: Exception) -> bool:
+        # We can not recover from formatting problems with the message.  Do not attempt to retry.
+        if isinstance(error, UnhandledMessage):
+            return False
+
+        return self.should_retry(record, error)
+
+    async def _handle_retry(self, record: ConsumerRecord, error: Exception) -> None:
+        MESSAGE_REQUEUED.labels(
+            stream_id=record.topic,
+            partition=record.partition,
+            error=error.__class__.__name__,
+            group_id=self._subscription.group,
+        ).inc()
+
+        await self.handle_retry(record, error)
+
+    async def _handle_failure(self, record: ConsumerRecord, error: Exception) -> None:
+        MESSAGE_FAILED.labels(
+            stream_id=record.topic,
+            partition=record.partition,
+            error=error.__class__.__name__,
+            group_id=self._subscription.group,
+        ).inc()
+
+        try:
+            await self.handle_failure(record, error)
+        except UnhandledMessage:
+            # Here we swallow any errors related to input format to mantain existing functionality
```

Comment:
Shouldn't we log here?

Reply:
Currently `kafkaesk.app.ConsumerSubscriber` is logging (and tracking metrics) on `UnhandledMessage` exceptions prior to passing those off to the policy.  I didn't want to duplicate those so I left them as they were and didn't log in the policy.

If you feel it makes more sense to move all of that logic into the policy that's easy enough.

Reply:
I wonder if we shouldn't even be catching this error here? Should it be possible for this to happen?

Reply:
Even though it is impossible to re-try messages that riase these exceptions (because we can not de-serialize the message) we still want them handled by the retry policies so they can be added to failure topics and reviewed later.  However, I do think I need to re-work this a little bit into a more generic "don't requeue these exceptions" type of system rather than hardcoding the behavior for this one single exception.

Reply:
Ok, I've adjusted this a little bit.  There is now a global `no_requeue_exceptions` list and each policy can also be configured with its own `no_requeue_exceptions` list.  If a policy receives these exceptions they will be sent directly to the failure handler.

Additionally, I have moved the logic to no re-raise `UnhandledMessage` exceptions back into the `SubscriptionConsumer`.  This is there to preserve the current functionality when the default `NoRetry` policy is set.